pkg/sql/pg_catalog.go

```diff
@@ -834,10 +832,243 @@ var (
+			conbin,         // conbin
+			consrc,         // consrc
+			condef,         // condef
+		); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type oneAtATimeSchemaResolver struct {
+	ctx context.Context
+	p   *planner
+}
+
+func (r oneAtATimeSchemaResolver) getDatabaseByID(id sqlbase.ID) (*DatabaseDescriptor, error) {
+	return r.p.Tables().databaseCache.getDatabaseDescByID(r.ctx, r.p.txn, id)
+}
+
+func (r oneAtATimeSchemaResolver) getTableByID(id sqlbase.ID) (*TableDescriptor, error) {
+	table, err := r.p.LookupTableByID(r.ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	return table.Desc.TableDesc(), nil
+}
+
+// makeAllRelationsPGCatalogTable creates a virtual table that searches through
+// all table descriptors in the system. It automatically adds an index to the
+// table id column as well. The input schema must have a single INDEX definition
+// with a single column, which must be the column that contains the table id.
+// indexesAreNotComplete should be set to true if the index constraint
+// might not be something that can be used to search by. For example, a common
+// case is having a table that's keyed by an OID that can either be an index
+// or a table OID. In this case, you must pass true for this variable, and if
+// the input is an index OID, the index won't be used. Another common case and
+// gotcha to look out for is that some tables do not have entries in
+// system.descriptor, such as other virtual tables.
+func makeAllRelationsPGCatalogTable(
+	comment string,
+	schemaDef string,
+	virtualOpts virtualOpts,
+	indexesAreNotComplete bool,
+	populateFromTable func(ctx context.Context, p *planner, h oidHasher, db *sqlbase.DatabaseDescriptor,
+		scName string, table *sqlbase.TableDescriptor, lookup simpleSchemaResolver,
+		addRow func(...tree.Datum) error,
+	) error,
+) virtualSchemaTable {
+	populateAll := func(ctx context.Context, p *planner, dbContext *DatabaseDescriptor, addRow func(...tree.Datum) error) error {
+		h := makeOidHasher()
+		return forEachTableDescWithTableLookup(ctx, p, dbContext, virtualOpts,
+			func(db *sqlbase.DatabaseDescriptor, scName string, table *sqlbase.TableDescriptor, lookup tableLookupFn) error {
+				return populateFromTable(ctx, p, h, db, scName, table, lookup, addRow)
+			})
+	}
+	return virtualSchemaTable{
+		comment: comment,
+		schema:  schemaDef,
+		indexes: []virtualIndex{
+			{
+				partial: indexesAreNotComplete,
+				populate: func(ctx context.Context, constraint tree.Datum, p *planner, db *DatabaseDescriptor,
+					addRow func(...tree.Datum) error) (bool, error) {
+					var id sqlbase.ID
+					switch t := constraint.(type) {
+					case *tree.DOid:
+						id = sqlbase.ID(t.DInt)
+					case *tree.DInt:
+						id = sqlbase.ID(*t)
+					default:
+						return false, errors.AssertionFailedf("unexpected type %T for table id column in virtual table %s",
+							constraint, schemaDef)
+					}
+					table, err := p.LookupTableByID(ctx, id)
+					if err != nil {
```

Comment:
you want to be checking if it is an UndefinedRelationErr before doing this -- it could be a wide variety of other errors that you're swallowing here.